python/prophet/forecaster.py

```diff
@@ -1438,19 +1482,160 @@ def predict_uncertainty(self, df):
+        if df['t'].max() <= 1:
+            # there is no trend uncertainty in historic trends
+            sample_trends = np.zeros((k, len(df)))
+        else:
+            future_df = df[df['t'] > 1]
+            # below calculates 't' if unavailable
+            # future_time_series = ((future_df["ds"] - self.start) / self.t_scale).values
+            n_length = len(future_df)
+            # handle 1 length futures by using history
+            if n_length > 1:
+                single_diff = np.diff(future_df['t']).mean()
+            else:
+                single_diff = np.diff(self.history['t']).mean()
+            change_likelihood = len(self.changepoints_t) * single_diff
+            deltas = self.params["delta"][0]
+            mean_delta = np.mean(np.abs(deltas)) + 1e-8
+            if self.growth == "linear":
+                mat = self._make_trend_shift_matrix(mean_delta, change_likelihood, n_length, k=k)
+                sample_trends = mat.cumsum(axis=1).cumsum(axis=1)  # from slope changes to actual values
+                sample_trends = sample_trends * single_diff  # scaled by the actual meaning of the slope
+            elif self.growth == "logistic":
+                mat = self._make_trend_shift_matrix(mean_delta, change_likelihood, n_length, k=k)
+                cap_scaled = (future_df["cap"] / self.y_scale).values
+                sample_trends = self._logistic_uncertainty(
+                    mat, deltas, cap_scaled,
+                    future_df['t'].to_numpy(), k=k,
+                    n_length=n_length, single_diff=single_diff,
+                )
+            elif self.growth == "flat":
+                # no trend uncertainty when there is no growth
+                sample_trends = np.zeros((k, n_length))
+            else:
+                raise NotImplementedError
+            # handle past included in dataframe
+            if df['t'].min() <= 1:
+                past_trend = np.zeros((k, np.sum(df['t'] <= 1)))
+                sample_trends = np.concatenate([past_trend, sample_trends], axis=1)
 
-        series = {}
-        for key in ['yhat', 'trend']:
-            series['{}_lower'.format(key)] = self.percentile(
-                sim_values[key], lower_p, axis=1)
-            series['{}_upper'.format(key)] = self.percentile(
-                sim_values[key], upper_p, axis=1)
+        # add gaussian noise based on historical levels
+        sigma = self.params["sigma_obs"][0]
+        historical_variance = np.random.normal(scale=sigma, size=sample_trends.shape)
+        return sample_trends, historical_variance
 
-        return pd.DataFrame(series)
+    @staticmethod
+    def _make_historical_mat_time(deltas, changepoints_t, t_time, n_row=1, single_diff=None):
+        """
+        Creates a matrix of slope-deltas where these changes occured in training data according to the trained prophet obj
+        """
+        if single_diff is None:
+            single_diff = np.diff(t_time).mean()
```

Comment:
is this really necessary? won't single_diff always be calculated before? we can make it non optional, and not pass t_time to this function at all

Reply:
that's a good point, it's totally unnecessary, but also the is None calculation is so fast it doesn't make much of a performance difference either way. I can remove it if other changes are being made.